docs/provider: Initial Maintainers Guide (#9332)
terraform-providers/terraform-provider-aws: .github/MAINTAINING.md

# Maintaining the Terraform AWS Provider

<!-- TOC depthFrom:2 -->

- [Pull Requests](#pull-requests)
    - [Pull Request Review Process](#pull-request-review-process)
        - [Dependency Updates](#dependency-updates)
            - [AWS Go SDK Updates](#aws-go-sdk-updates)
            - [Terraform Updates](#terraform-updates)
    - [Pull Request Merge Process](#pull-request-merge-process)
- [Release Process](#release-process)

<!-- /TOC -->

## Pull Requests

### Pull Request Review Process

Notes for each type of pull request are (or will be) available in subsections below.

> - If you are planning on owning the pull request through merge/closure, assign it to yourself

Reviewer: By "through" do you mean the duration (until it's merged/closed) or the action or doing so?

Author: The intention here was taking leadership of the pull request until its merged/closed or reassigned. More importantly, a protection mechanism against multiple people working on the same pull request.

Reviewer: Makes sense! What about something like "If you plan to be responsible for the pull request, through to the merge/close process ..."?

Author: Sounds good. 👍

- Add `bug`, `enhancement`, `new-data-source`, `new-resource`, or `technical-debt` labels to match expectations from change
- Review the contents of the pull request and ensure the change follows the relevant section of the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#checklists-for-contribution)
- If the change is not acceptable, leave a long form comment about the reasoning and close the pull request
- If the change is acceptable with modifications, leave a pull request review marked using the `Request Changes` option (for maintainer pull requests with minor modification requests, giving feedback with the `Approve` option is recommended so they do not need to wait for another round of review)

> - If the author is unresponsive for changes (generally we give two weeks), determine importance and level of effort to finish the pull request yourself including their commits or close the pull request

Reviewer: ```suggestion
- If the author is unresponsive for changes (by default we give two weeks), determine importance and level of effort to finish the pull request yourself including their commits or close the pull request
```

> - Run relevant acceptance testing (locally or in TeamCity) and ensure no new failures are being introduced

Reviewer: Can this link to any additional documentation on running acceptance tests? When does it make sense to use TeamCity?

Reviewer: Also -- where does a maintainer look to decide which regions or clouds the tests should be run against?

Author: Hmm. We can likely link to the [Contributing Guide section for running acceptance testing](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#running-an-acceptance-test). The TeamCity documentation and access to it are considered internal at the moment. I'm not sure we have plans to change either of those. TeamCity is generally helpful when there is a lot to test or those tests will potentially take a long time. The heuristics for those are fuzzy.

Author: Updated this line to include link to Contributing Guide as well as mentioning AWS Commercial and GovCloud (US) although GovCloud (US) is not a strict requirement at the moment.

- Approve the pull request with a comment outlining what steps you took that ensure the change is acceptable, e.g. acceptance testing output

``````markdown
Looks good, thanks @username! :rocket:

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAcc...
--- PASS: TestAcc...
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAcc...
--- PASS: TestAcc...
```
``````

#### Dependency Updates

##### AWS Go SDK Updates

> Almost exclusively, `github.com/aws/aws-sdk-go` updates are additive in nature and it is generally safe to only scan through them before approving and merging. If you have any concerns about any of the service client updates such as suspicous code removals in the update or deprecations introduced, run the acceptance testing for potentially affected resources before merging.

Reviewer: ```suggestion
Almost exclusively, `github.com/aws/aws-sdk-go` updates are additive in nature. It is generally safe to only scan through them before approving and merging. If you have any concerns about any of the service client updates such as suspicious code removals in the update, or deprecations introduced, run the acceptance testing for potentially affected resources before merging.
```

What would a new deprecation look like?

Other important notes:

> - CloudFront service client updates have previously caused an issue when a new field introduced in the SDK was not included with Terraform and caused all requests to error (https://github.com/terraform-providers/terraform-provider-aws/issues/4091). As a precaution, if you see CloudFront updates, run all the CloudFront resource acceptance testing before merging.

Reviewer: ```suggestion
- CloudFront service client updates have previously caused an issue when a new field introduced in the SDK was not included with Terraform and caused all requests to error (https://github.com/terraform-providers/terraform-provider-aws/issues/4091). As a precaution, if you see CloudFront updates, run all the CloudFront resource acceptance testing before merging (`TestAccAWSCloudFront`).
```

##### Terraform Updates

Run the full acceptance testing suite against the pull request and verify there are no new or unexpected failures.

### Pull Request Merge Process

- Add this pull request to the upcoming release milestone
- Add any linked issues that will be closed by the pull request to the same upcoming release milestone
- Merge the pull request
- Delete the branch (if the branch is on this repository)

> - Update the repository `CHANGELOG.md`. See also the [Extending Terraform documentation](https://www.terraform.io/docs/extend/best-practices/versioning.html) for more information about the expected CHANGELOG format.

Reviewer: ```suggestion
- Update the repository `CHANGELOG.md` by directly committing to the `master` branch. See also the [Extending Terraform documentation](https://www.terraform.io/docs/extend/best-practices/versioning.html) for more information about the expected CHANGELOG format.
```

- Leave a comment on any issues closed by the pull request noting that it has been merged and when to expect the release containing it, e.g.

```markdown

> The fix for this has been merged and will release with version X.Y.Z of the Terraform AWS Provider, likely in the XXX timeframe.

Reviewer: ```suggestion
The fix for this has been merged and will release with version X.Y.Z of the Terraform AWS Provider, expected in the XXX timeframe.
```

"Scheduled" could also work -- "likely" sounds kind of indecisive.

```

## Release Process

- Ensure a milestone exists for the next release after this release (Generally, the next milestone will be a minor version increase unless previously decided for a major or patch version)

> - Check the upcoming release milestone for open items and either work through them or move them to the next milestone

Reviewer: The sequence feels a little odd to me -- maybe it could be something like "Check the upcoming release milestone for open items, create the next milestone if it doesn't exist yet, finish things or move them"?

Author: The idea here is to have the next milestone pre-created so maintainers working later on do not need to worry about asynchronously creating it. I'll adjust the wording of the line above to try and help clarify. 👍

- Run the HashiCorp (non-OSS) TeamCity release job with the `DEPLOYMENT_TARGET_VERSION` matching the expected release milestone and `DEPLOYMENT_NEXT_VERSION` matching the next release milestone

> - Wait for the TeamCity release job and TeamCity website deployment jobs to complete

Reviewer: Is there an email or other notification?

Author: Yes, there is a start and finish notification in our internal Slack.
